src/HonzaBotner.Discord.Services/EventHandlers/BadgeRoleHandler.cs

```diff
@@ -0,0 +1,73 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using DSharpPlus.Entities;
+using DSharpPlus.EventArgs;
+using HonzaBotner.Discord.EventHandler;
+using HonzaBotner.Discord.Services.Options;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace HonzaBotner.Discord.Services.EventHandlers;
+
+public class BadgeRoleHandler
+{
+    private readonly ILogger<BadgeRoleHandler> _logger;
+    private readonly BadgeRoleOptions _options;
+
+    public BadgeRoleHandler(IOptions<BadgeRoleOptions> options, ILogger<BadgeRoleHandler> logger)
+    {
+        _options = options.Value;
+        _logger = logger;
+    }
+
+    public async Task<EventHandlerResult> Handle(GuildMemberUpdateEventArgs args)
+    {
+        await Task.Delay(0); // Avoid warning that there is no "await" in async method
+        if (args.RolesAfter.Count > args.RolesBefore.Count) await CheckAddedRoles(args);
+        else if (args.RolesAfter.Count < args.RolesBefore.Count) await CheckRemovedRoles(args);
+        return EventHandlerResult.Continue;
+    }
+
+    private async Task CheckAddedRoles(GuildMemberUpdateEventArgs args)
+    {
+        DiscordRole addedRole = args.RolesAfter.First(role => !args.RolesBefore.Contains(role));
+        if (_options.PairedRoles.ContainsKey(addedRole.Id))
```

Comment:
Not checking if the Member has trigger role - should pass this if only with trigger role